pkg/distributor/ha_tracker.go

```diff
@@ -0,0 +1,316 @@
+package distributor
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/weaveworks/common/mtime"
+
+	"github.com/cortexproject/cortex/pkg/ingester/client"
+	"github.com/cortexproject/cortex/pkg/ring"
+	"github.com/cortexproject/cortex/pkg/util"
+)
+
+const (
+	longPollDuration = 10 * time.Second
+)
+
+// ErrDontOverwrite is used when we want to accept a sample but not the value in consul when CAS'ing.
+type ErrDontOverwrite struct {
+}
+
+func (e *ErrDontOverwrite) Error() string {
+	return fmt.Sprintf("don't overwrite")
+}
+
+// Track the most recent timestamp for the replica we're
+// accepting samples from in a Prometheus HA setup.
+type replicaTracker struct {
+	replica   string
+	timestamp int64
+}
+
+// Track the replica we're accepting samples from
+// for each HA cluster we know about.
+type tracker struct {
+	logger log.Logger
+
+	// Replicas we are accepting samples from.
+	electedLock sync.RWMutex
+	elected     map[string]replicaTracker
+	done        chan struct{}
+	quit        context.CancelFunc
+	// The labels to look for as bytes.
+	clusterBytes []byte
+	replicaBytes []byte
+	prefix       string
+
+	// We should only update the timestamp this often.
+	writeTimeout int64
```

Comment:
Why not `time.Duration`?

Reply:
Timestamps within samples are int64. We could import the Prometheus timestamp package and convert back to a time type?

Reply:
Oh I see.  Lets keep these a time.Durations, and do the conversion on the Prometheus timestamps.

Note Prometheus timestamps are milliseconds, not nano.